core/changedetector/proto_change_detector_service.go

```diff
@@ -0,0 +1,215 @@
+package changedetector
+
+import (
+	"context"
+	"errors"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/google/uuid"
+	"github.com/goto/stencil/pkg/newrelic"
+	stencilv1beta2 "github.com/goto/stencil/proto/gotocompany/stencil/v1beta1"
+	"google.golang.org/protobuf/types/descriptorpb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	"log"
+	"time"
+)
+
+func NewService(nr newrelic.Service) *Service {
+	return &Service{
+		newrelic: nr,
+	}
+}
+
+type Service struct {
+	newrelic newrelic.Service
+}
+
+func (s *Service) IdentifySchemaChange(ctx context.Context, request *ChangeRequest) (*stencilv1beta2.SchemaChangedEvent, error) {
+	endFunc := s.newrelic.StartGenericSegment(ctx, "Identify Schema Change")
+	defer endFunc()
+	sce := &stencilv1beta2.SchemaChangedEvent{
+		EventId:        uuid.New().String(),
+		EventTimestamp: timestamppb.New(time.Now()),
+		NamespaceName:  request.NamespaceName,
+		SchemaName:     request.SchemaName,
+		Version:        request.Version,
+	}
+	if request.OldData == nil || len(request.OldData) == 0 {
+		log.Println("IdentifySchemaChange failed as previous schema data is nil")
+		return nil, errors.New("previous schema data is nil")
+	}
+	currentFds, err := getDescriptorSet(request.NewData)
+	if err != nil {
+		log.Printf("unable getSchemaChangeEvent get file descriptor set from current schema data %v\n", err)
+		return nil, errors.New("unable getSchemaChangeEvent get file descriptor set from current schema data")
+	}
+	prevFds, err := getDescriptorSet(request.OldData)
+	if err != nil {
+		log.Printf("unable getSchemaChangeEvent get file descriptor set from previous schema data %v\n", err)
+		return nil, errors.New("unable getSchemaChangeEvent get file descriptor set from previous schema data")
+	}
+	setDirectlyImpactedSchemasAndFields(currentFds, prevFds, sce)
+	reverseDependencies := make(map[string][]string)
+	populateReverseDependencies(currentFds, reverseDependencies)
+	for _, schema := range sce.UpdatedSchemas {
+		appendImpactedDependents(sce, schema, findDependentImpactedSchemas(reverseDependencies, schema))
+	}
+	return sce, nil
+}
+
+func appendImpactedDependents(sce *stencilv1beta2.SchemaChangedEvent, key string, impactedDependents []string) {
+	if sce.ImpactedSchemas == nil {
+		sce.ImpactedSchemas = make(map[string]*stencilv1beta2.ImpactedSchemas)
+	}
+	sce.ImpactedSchemas[key] = &stencilv1beta2.ImpactedSchemas{
+		SchemaNames: impactedDependents,
+	}
+}
+
+func setDirectlyImpactedSchemasAndFields(currentFds, prevFds *descriptor.FileDescriptorSet, sce *stencilv1beta2.SchemaChangedEvent) {
```

Comment:
Since, I am passing the reference any change would be reflected in the original object. Hence, comparing the whole objects while unit testing. Also, with this all updates are happening in the same object.